sympy/calculus/tests/test_hamilton.py

```diff
@@ -0,0 +1,87 @@
+from sympy import Symbol, Function, Derivative as D, Eq, cos, sin
+from sympy.utilities.pytest import raises
+from sympy.calculus.hamilton import hamilton_equations as hamilton
+
+
+def test_hamilton_errors():
+    # Test by giving the Hamiltonian Interface bad-functions
+
+    # ERROR CASE: 1
+    # =============
+    x = Function('x')
+    p = Function('p')
+    t = Symbol('t')
+    h1 = x(t)**2 / 2 + p(t)**2 / 2
+    h2 = x(p)**2 / 2 + p(t)**2 / 2
+    raises(TypeError, lambda: hamilton(h1, x(t), p, t))
+    raises(TypeError, lambda: hamilton(h2, x(t), p(t), t))
+
+    # ERROR CASE: 2
+    # =============
+    x = Function('x')
+    p = Function('p')
+    t = Symbol('t')
+    h = x(t) ** 2 / 2 + p(t) ** 2 / 2
+    # Giving function instead of symbol as time
+    raises(TypeError, lambda: hamilton(h, [x(t)], [p(t)], x(t)))
+
+    # ERROR CASE: 3
+    # =============
+    x = Function('x')
+    p = Function('p')
+    k = Symbol('k')
+    t = Symbol('t')
+    h = x(k) ** 2 / 2 + p(k) ** 2 / 2
+    # Giving wrong dependent variable
+    raises(ValueError, lambda: hamilton(h, [x(k)], [p(k)], t))
+
+    # ERROR CASE: 4
+    # =============
+    x = Symbol('x')
+    p = Symbol('p')
+    t = Symbol('t')
+    h1 = x ** 2 / 2 + p ** 2 / 2
+    # Giving symbol instead of function as the Hamiltonian
+    raises(TypeError, lambda: hamilton(h1, [x], [p], t))
+
+
+def test_hamilton_shm1():
+    # Test by solving for SHM in 1-Dimension
+    x = Function('x')
+    p = Function('p')
+    t = Symbol('t')
+    h1 = x(t)**2 / 2 + p(t)**2 / 2
+    assert hamilton(h1, x(t), p(t), t) == [
+        Eq(x(t) + D(p(t), t), 0), Eq(p(t) - D(x(t), t), 0)
+    ]
+
+
+def test_hamilton_shm2():
+    # Test by solving for SHM in 2-Dimensions
+    x = Function('x')
+    p = Function('p')
+    t = Symbol('t')
+    y = Function('y')
+    q = Function('q')
+    h2 = x(t)**2 / 2 + y(t)**2 / 2 + p(t)**2 / 2 + q(t)**2 / 2
+    assert hamilton(h2, [x(t), y(t)], [p(t), q(t)], t) == [
+        Eq(x(t) + D(p(t), t), 0), Eq(p(t) - D(x(t), t), 0),
+        Eq(y(t) + D(q(t), t), 0), Eq(q(t) - D(y(t), t), 0)
+    ]
+
+
+def test_hamilton_shm3():
+    # Test by solving for SHM over a cylinder
+    t = Symbol('t')
+    qr = Function('R')
+    qt = Function('Theta')
+    qz = Function('Z')
+    pr = Function('pR')
+    pt = Function('pTheta')
+    pz = Function('pZ')
+    h3 = pt(t)**2/2 + pz(t)**2/2 + qz(t)**2/2
+    assert hamilton(h3, [qr(t), qt(t), qz(t)], [pr(t), pt(t), pz(t)], t) == [
```

Comment:
I imagine that most people would create the Functions like symbols and then use them in an expression. We could eliminate a lot of redundancy by allowing entry like `(h3, (qr, qt, qz), (pr, pt, pz), t)` and then recompose them in the function: 
```python
for q, p in funcs:
    coord, momentum = q(t), p(t)
...
```

Reply:
(But if the arguments would have be more than `t` then this wouldn't work.)

Reply:
Yeah, but won't it just be better to leave it as is, since it's both similar to it's Lagrangian counterpart, and as you pointed out, could potentially have more arguments than t. Tell me what I should do @smichr.